api/openai/chat.go

```diff
@@ -298,102 +368,134 @@ func ChatEndpoint(cm *config.ConfigLoader, o *options.Option) func(c *fiber.Ctx)
+
+						// If there is a message that the LLM already sends as part of the JSON reply, use it
+						arguments := map[string]interface{}{}
+						json.Unmarshal([]byte(args), &arguments)
+						m, exists := arguments["message"]
+						if exists {
+							switch message := m.(type) {
+							case string:
+								if message != "" {
+									log.Debug().Msgf("Reply received from LLM: %s", message)
+									message = backend.Finetune(*config, predInput, message)
+									log.Debug().Msgf("Reply received from LLM(finetuned): %s", message)
+
+									*c = append(*c, schema.Choice{Message: &schema.Message{Role: "assistant", Content: &message}})
+									return
+								}
 							}
 						}
-					}
 
-					log.Debug().Msgf("No action received from LLM, without a message, computing a reply")
-					// Otherwise ask the LLM to understand the JSON output and the context, and return a message
-					// Note: This costs (in term of CPU) another computation
-					config.Grammar = ""
-					images := []string{}
-					for _, m := range input.Messages {
-						images = append(images, m.StringImages...)
-					}
-					predFunc, err := backend.ModelInference(input.Context, predInput, images, o.Loader, *config, o, nil)
-					if err != nil {
-						log.Error().Msgf("inference error: %s", err.Error())
-						return
-					}
+						log.Debug().Msgf("No action received from LLM, without a message, computing a reply")
+						// Otherwise ask the LLM to understand the JSON output and the context, and return a message
+						// Note: This costs (in term of CPU) another computation
+						config.Grammar = ""
+						images := []string{}
+						for _, m := range input.Messages {
+							images = append(images, m.StringImages...)
+						}
+						predFunc, err := backend.ModelInference(input.Context, predInput, images, o.Loader, *config, o, nil)
+						if err != nil {
+							log.Error().Msgf("inference error: %s", err.Error())
+							return
+						}
 
-					prediction, err := predFunc()
-					if err != nil {
-						log.Error().Msgf("inference error: %s", err.Error())
-						return
+						prediction, err := predFunc()
+						if err != nil {
+							log.Error().Msgf("inference error: %s", err.Error())
+							return
+						}
+
+						fineTunedResponse := backend.Finetune(*config, predInput, prediction.Response)
+						*c = append(*c, schema.Choice{Message: &schema.Message{Role: "assistant", Content: &fineTunedResponse}})
+					} else {
+						if len(input.Tools) > 0 {
+							// Result is different in the case we have a tool call
+							*c = append(*c, schema.Choice{
+								FinishReason: "tool_calls",
+								Message: &schema.Message{
+									Role: "assistant",
+									ToolCalls: []schema.ToolCall{
+										{
+											ID:   id,
+											Type: "function",
+											FunctionCall: schema.FunctionCall{
+												Name:      name,
+												Arguments: args,
+											},
+										},
+									},
+									FunctionCall: ss,
```

Comment:
Is it correct that this is also in here in case of tool calls?

Reply:
good catch! copy-pasta error :)